internal/pkg/etcd/etcd.go

```diff
@@ -102,12 +108,83 @@ func (c *Client) ValidateForUpgrade(ctx context.Context, config config.Provider,
 			if len(member.Name) == 0 {
 				return fmt.Errorf("etcd member %d is not started, all members must be running to perform an upgrade", member.ID)
 			}
+
+			if err = validateMemberHealth(ctx, member.GetPeerURLs()); err != nil {
+				return fmt.Errorf("etcd member %d is not healthy; all members must be health to perform an upgrade: %w", member.ID, err)
+			}
+		}
+
+		// Member count must not be an even number, lest the uprgade cause a loss of quorum.
+		if len(resp.Members)%2 == 0 {
+			return fmt.Errorf("etcd member count (%d) is even; upgrade of this node may cause loss of quorum", len(resp.Members))
 		}
+
+		// Quorum must be good before allowing an upgrade of an etcd member.
+		if err = c.ValidateQuorum(ctx); err != nil {
+			return fmt.Errorf("etcd membership quorum validation failed: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// ValidateQuorum performs a KV operation to make certain that quorum is good.
+func (c *Client) ValidateQuorum(ctx context.Context) (err error) {
+	// Get a random key. As long as we can get the response without an error, quorum is good.
+	checkCtx, cancel := context.WithTimeout(ctx, QuorumCheckTimeout)
+	defer cancel()
+
+	_, err = c.Get(checkCtx, "health")
+	if err == rpctypes.ErrPermissionDenied {
+		// Permission denied is OK since proposal goes through consensus to get this error.
+		err = nil
+	}
+
+	if err != nil {
+		return err
 	}
 
 	return nil
 }
 
+func validateMemberHealth(ctx context.Context, memberURIs []string) error {
+	type healthResult struct {
+		Health string `json:"health"`
+		Reason string `json:"reason"`
+	}
+
+	for _, u := range memberURIs {
+		uri, err := url.Parse(u)
+		if err != nil {
+			continue
+		}
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("http://%s/health", uri.Host), nil)
```

Comment:
I don't know what exactly does this method check in terms of etcd health status (looks like many ways to check health!), but if we follow etcd recommended way, we should check for etcd to be available to commit quorum on every node (including our local node). 

this might be achieved by checking the health of all `service[etcd]` on Talos nodes

as we're inside the cluster here, we can use following trick to build Talos API client: 

https://github.com/talos-systems/talos/blob/master/pkg/cluster/local.go#L34

this client will be fully featured though (can proxy to other nodes, etc.)

with that we could do `client.ServiceInfo(client.WithNodes(ctx, masterNodes...), "etcd")` and parse response like we do in https://github.com/talos-systems/talos/blob/master/pkg/cluster/check/service.go#L129

Reply:
I'm a bit confused on this. Wouldn't this check be only as good as the last time the health check was performed? Wouldn't we need a nearly instant check just before upgrading?

Reply:
In other words, isn't our accuracy is influenced by how long ago the health check was run?